Update Homekit climate.py to remap current mode

GaryOkie opened the pull request:
This update changes the mapping of Homekit's Current Mode Heating/Cooling State to show the HASS hvac_action attribute as "idle" instead of "off" when the returned value is 0.

![Homekit_Tstat](https://user-images.githubusercontent.com/37629938/68430245-5493df80-0175-11ea-9cdb-53fa443a391a.JPG)

```
entity:  climate.icomfort_s30
operation:    Heat
hvac_modes: off,heat,cool,heat_cool
current_temperature: 57
min_temp: 40
max_temp: 90
temperature: 55
current_humidity: 57
hvac_action: idle
friendly_name: Downstairs Thermostat
supported_features: 1
```

## Breaking Change:

Thermostats supported by the Homekit Controller component were reporting "**off**" for the running status when the HVAC was powered on but not actively heating or cooling.   This _hvac_action_ attribute will now properly return "**idle**" instead of off.    Any logic that tests for the hvac_action condition of off will need to change to idle.   There is no change to any other states.


**Related issue (if applicable):** related to discussion in  issue #20885

**Pull request with documentation for [home-assistant.io](https://github.com/home-assistant/home-assistant.io) (if applicable):** home-assistant/home-assistant.io#<home-assistant.io PR number goes here>

## Example entry for `configuration.yaml` (if applicable):
```yaml

```

## Checklist:
  - [x ] The code change is tested and works locally 
  - [ ] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.
  - [ ] I have followed the [development checklist][dev-checklist]

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [ ] [_The manifest file_][manifest-docs] has all fields filled out correctly. Update and include derived files by running `python3 -m script.hassfest`.
  - [ ] New or updated dependencies have been added to `requirements_all.txt` by running `python3 -m script.gen_requirements_all`.
  - [ ] Untested files have been added to `.coveragerc`.

If the code does not interact with devices:
  - [ ] Tests have been added to verify that the new code works.

[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html

GaryOkie committed:
Update Homekit climate.py to remap current mode

This update changes the mapping of Homekit's Current Mode Heating/Cooling State  to show the HASS Hvac_action attribute as "idle" instead of "off" when the returned value is 0.

homeassistant commented:
Hi @GaryOkie, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#28625).

Once you do that we will be able to review and accept this pull request.

Thanks!

Jc2k commented on homeassistant/components/homekit_controller/climate.py:
```diff
@@ -14,6 +14,7 @@
     HVAC_MODE_HEAT,
     HVAC_MODE_OFF,
     CURRENT_HVAC_OFF,
```
`CURRENT_HVAC_OFF` isn't used any more so you will probably get told off by the commit bots for not removing it from the imports.

GaryOkie replied:
all new territory for me, navigating github PR processes.   So it looks like you already removed CURRENT_HVAC_OFF.   Is there anything else I need to do?

Jc2k replied:
I haven't removed it so you'll need to

GaryOkie replied:
I'm sorry @jc2k, but I am struggling to find my branch containing climate.py that I can edit.

GaryOkie replied:
I found this version, but it is very different.  I guess from an HA repo copied quite a while ago? https://github.com/GaryOkie/home-assistant/blob/dev/homeassistant/components/homekit_controller/climate.py

GaryOkie replied:
Duh, just found the secret menu to edit the dang file.   Hidden under the "..." > Edit.    Sheesh, sorry my noobness.

Jc2k commented:
Just had a quick look at the tests [here](https://github.com/home-assistant/home-assistant/blob/dev/tests/components/homekit_controller/test_climate.py) - theres an `off` [here](https://github.com/home-assistant/home-assistant/blob/dev/tests/components/homekit_controller/test_climate.py#L216) that might need changing as well.

GaryOkie committed:
Update climate.py

removed imported constant no longer being used  (CURRENT_HVAC_OFF)

GaryOkie reviewed (commented):
removed unused constant CURRENT_HVAC_OFF

GaryOkie committed:
corrected update to climate.py

trying again to remove unused constant.

GaryOkie reviewed (commented):
trying again to remove imported unused constant (CURRENT_HVAC_OFF)

GaryOkie reviewed (commented):
updated test_climate.py (all checks passed), so hoping these tests pass for climate.py now as well.

GaryOkie commented:
I thought that by doing a new PR to update test_climate.py to change the assert hvac_action from off to idle, that these errors would be cleared.    That PR passed, but that alone did not solve this.   Somewhere else it's picking up the original test code I guess.      Sorry, but I have not found out how to make these tests pass.

GaryOkie committed:
Update test_climate.py

Jc2k commented:
Hey sorry was afk but it looks like you figured it out and everything is happy and passing. Everything looks good here - I would say that the comment you added to the test isn’t needed. It’s best not to refer to changes like that - we will be able to see this commit in ‘git annotate’. Your comments show be more in the present - you could say “Ensure we handle when the device is in heating mode but idle”. The fact that it’s a change is meaningless the second we merge it.

Other than that, well done. If you can edit that then I can approve it tonight / tomorrow

GaryOkie committed:
removed "change" comment

The added comment describing the change was not needed and should not be included, as it will already be described via "git annotate"  (per @jc2k)

GaryOkie reviewed (commented):
removed unnecessary "change" comment

GaryOkie commented:
What the heck?!   Removing a comment in test_climate.py somehow leads to a new error with a xiaomi vacuum (test_xiaomi_exceptions.py)???

Jc2k commented:
Flaky tests happen some time - don’t worry about it. A dev can request a rebuild if it happens - I’ll see if I have permission when I’m at my desk. If a test is flaky we disable it or get it’s dev to fix it but some of them are really rare failures

GaryOkie commented:
So the test_climate test is run along with the included test code for every component.  If any of those tests has an error, the check will fail.  The developer of the test that failed needs to be contacted to fix it before I can run the check again?   

Is that what we're waiting on now?   Just trying to better understand this HA/Git PR check process!

Jc2k commented:
So yes, every change gets the whole test suite run against it. It could be that the homekit tests somehow rely on homekit_controller to run an integration test (they don't) and this change actually broke their tests. The only way to be sure is to run all the tests every time.

But no, that doesn't neccesarily mean you are blocked and waiting for the other developer. As it's a flaky test running the tests again may actually get it to go green. In reality you were waiting for me to be near a computer and remember that home assistant was a thing 😊 (very busy at the moment).

The checks are re-running now.

Jc2k reviewed (approved):
If the tests pass then this is ready to merge 👍

MartinHjelmare commented:
This is a breaking change, right?

GaryOkie commented:
I'll defer to @jc2k as I'm new at this, but yeah, it's possible that someone could be using "hvac_action" in an automation and testing for "off", which would now (properly) state "idle".     More likely, an HVAC automation would be testing for "heating" or "cooling" to say, calculate energy use/run times, not off or idle times, but it's possible.

MartinHjelmare commented:
Please add a paragraph under the breaking change header in the PR description. Briefly describe what has changed and what the user needs to do to cope with the breaking change, as you've already outlined above.

GaryOkie commented:
OK, will do.   Sorry that it never occurred to me before until you asked that this could possibly break an automation.   It just seemed like a simple semantics update initially!

Jc2k commented:
Good call as ever @MartinHjelmare - had just seen it as a simple bug fix but you are right.

And thank you @GaryOkie for the quick turn around on the breaking change text. Looks like a good description of the change to me :+1:

MartinHjelmare reviewed (approved):
Great!
